lib/child_process.js

```diff
@@ -176,17 +174,26 @@ exports.execFile = function(file /*, args, options, callback*/) {
 
     if (!callback) return;
 
-    // merge chunks
-    var stdout;
-    var stderr;
-    if (!encoding) {
-      stdout = Buffer.concat(_stdout);
-      stderr = Buffer.concat(_stderr);
-    } else {
-      stdout = _stdout;
-      stderr = _stderr;
+    var stdout = Buffer.concat(_stdout, stdoutLen);
+    var stderr = Buffer.concat(_stderr, stderrLen);
+
+    var stdoutEncoding = encoding;
+    var stderrEncoding = encoding;
+
+    if (stdoutState.decoder) {
+      stdoutEncoding = stdoutState.decoder.encoding;
     }
 
+    if (stderrState.decoder) {
```

Comment:
minor nit: for consistency, these braces should be dropped or the other two `if`s below should have braces